cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+extern crate libc;
+extern crate soroban_env_host;
+
+use soroban_env_host::budget::Budget;
+use soroban_env_host::storage::{self, AccessType, SnapshotSource, Storage};
+use soroban_env_host::xdr::ScUnknownErrorCode::{General, Xdr};
+use soroban_env_host::xdr::{
+    self, AccountId, HostFunction, LedgerEntry, LedgerKey, ReadXdr, ScHostStorageErrorCode,
+    ScStatus, WriteXdr,
+};
+use soroban_env_host::{Host, HostError, LedgerInfo};
+use std::convert::TryInto;
+use std::error;
+use std::ffi::{CStr, CString};
+use std::panic;
+use std::ptr::null_mut;
+use std::rc::Rc;
+use xdr::LedgerFootprint;
+
+// TODO: we may want to pass callbacks instead of using global functions
+extern "C" {
+    // LedgerKey XDR in base64 string to LedgerEntry XDR in base64 string
+    fn SnapshotSourceGet(
+        handle: libc::uintptr_t,
+        ledger_key: *const libc::c_char,
+    ) -> *const libc::c_char;
+    // LedgerKey XDR in base64 string to bool
+    fn SnapshotSourceHas(handle: libc::uintptr_t, ledger_key: *const libc::c_char) -> libc::c_int;
+    // Free Strings returned from Go functions
+    fn FreeGoCString(str: *const libc::c_char);
+}
+
+struct CSnapshotSource {
+    handle: libc::uintptr_t,
+}
+
+impl SnapshotSource for CSnapshotSource {
+    fn get(&self, key: &LedgerKey) -> Result<LedgerEntry, HostError> {
+        let key_xdr = key
+            .to_xdr_base64()
+            .map_err(|_| ScStatus::UnknownError(Xdr))?;
+        let key_cstr = CString::new(key_xdr).map_err(|_| ScStatus::UnknownError(General))?;
+        let res = unsafe { SnapshotSourceGet(self.handle, key_cstr.as_ptr()) };
```

Comment:
dealing with multiple allocators is quite error prone. In fact, I don't know if this usage would work reliably, since GO might decide to GC the string ( since there are no more references to it ).
I think that the better solution here is to avoid mixing allocators altogether : 
make SnapshotSourceGet(C.uintptr_t, char*, void* val, C.uintptr_t in_val_len)
when called with  val == null and in_val_len == 0, the SnapshotSourceGet would return the required size.
then, a second call to SnapshotSourceGet with a staging buffer and a size would allow the SnapshotSourceGet to fill out that buffer.
This would also work around the need of base64 encoding.

Reply:
>  I don't know if this usage would work reliably, since GO might decide to GC the string ( since there are no more references to it ).

It works reliably because we are calling `C.CString()` on the Go side to allocate it. That's the idiomatic way to pass a string to C from Go. In case you haven't, I suggest you take a look at how to work with CGO.

>  think that the better solution here is to avoid mixing allocators altogether  ...

We are not mixing allocators, due to what I mentioned above.

Using a buffer+length is possible (CGO provides `C.CBytes` and `C.GoBytes`) but it's inconvenient due to bookeeping the length (see my answers to @leighmcculloch 's comments). If base64-encoding ends up being a bottleneck I will be happy to optimize that.

Reply:
This is tracked at https://github.com/stellar/soroban-tools/issues/367

Reply:
> Is it possible to use CString with Go byte slices converted to a string? The best of both worlds: we don't have to manage length, we don't have to do base64 enc/dec.

In general, you can’t. The problem is that CStrings are ended in zero and thus, you cannot represent byte slices containing zeros as a CString.

In particular, you cannot represent binary-marshaled XDR as a CString.